src/node/node_class.hpp

```diff
@@ -37,256 +57,1269 @@ using IndexPropertyType = js::IndexPropertyType<Types>;
+		ObjectSetPrototypeOf.Call({ ctor, parentCtor });
+	}
+
+	//use PropertyDescriptors instead of ClassPropertyDescriptors, since ClassPropertyDescriptors requires the JS instance members callbacks to be instance members of the WrappedObject<ClassType> class
+	if (isRealmObjectClass) {
+		std::vector<Napi::PropertyDescriptor> properties;
+		auto ctorPrototype = ctor.Get("prototype").As<Napi::Object>();
+		for (auto& pair : s_class.methods) {
+			auto descriptor = Napi::PropertyDescriptor::Function(env, ctorPrototype, Napi::String::New(env, pair.first) /*name*/, &method_callback, napi_default | realm::js::PropertyAttributes::DontEnum, (void*)pair.second /*callback*/);
+			properties.push_back(descriptor);
+		}
+
+		for (auto& pair : s_class.properties) {
+			napi_property_attributes napi_attributes = napi_default | (realm::js::PropertyAttributes::DontEnum | realm::js::PropertyAttributes::DontDelete);
+			auto descriptor = Napi::PropertyDescriptor::Accessor<property_getter_callback>(Napi::String::New(env, pair.first) /*name*/, napi_attributes, (void*)&pair.second /*callback*/);
+			properties.push_back(descriptor);
+		}
+		
+		ctorPrototype.DefineProperties(properties);
+	}
+
+	return ctor;
+}
+
+template<typename ClassType>
+Napi::Object ObjectWrap<ClassType>::create_instance(Napi::Env env, Internal* internal) {
+	Napi::EscapableHandleScope scope(env);
+
+
+	bool isRealmObjectClass = std::is_same<ClassType, realm::js::RealmObjectClass<realm::node::Types>>::value;
+
+	if (isRealmObjectClass && !internal) {
+		throw Napi::Error::New(env, "RealmObjectClass requires an internal realm object when creating instances");
+	}
+
+	Napi::Object factory = WrappedObject<ClassType>::create_instance(env, internal);
+    return scope.Escape(factory).As<Napi::Object>();
+}
+
+inline static void schema_object_type_constructor(const Napi::CallbackInfo& info) {
+}
+
+template<typename ClassType>
+void ObjectWrap<ClassType>::internalFinalizer(Napi::Env, typename ClassType::Internal* internal) {
+	delete internal;
+}
+
+static inline void remove_schema_object(std::unordered_map<std::string, SchemaObjectType*>* schemaObjects, const std::string& schemaName) {
+	bool schemaExists = schemaObjects->count(schemaName);
+	if (!schemaExists) {
+		return;
+	}
+
+	SchemaObjectType* schemaObjectType = schemaObjects->at(schemaName);
+	schemaObjects->erase(schemaName);
+	schemaObjectType->constructor.Reset();
+	delete schemaObjectType;
+}
+
+template<typename ClassType>
+Napi::Object ObjectWrap<ClassType>::create_instance_by_schema(Napi::Env env, Napi::Function& constructor, const realm::ObjectSchema& schema, Internal* internal) {
+	Napi::EscapableHandleScope scope(env);
+
+	bool isRealmObjectClass = std::is_same<ClassType, realm::js::RealmObjectClass<realm::node::Types>>::value;
+	if (!isRealmObjectClass) {
+		throw Napi::Error::New(env, "Creating instances by schema is supported for RealmObjectClass only");
```

Comment:
This tells me that we might be using the abstraction wrong - when a part of the API is only available for a subset of the users. Perhaps the creation of classes per object schema should be moved up the hierarchy?

Reply:
I don't see how you get to this conclusion. This is a way to switch method implementation based on the template parameter. Another way is template specialization which is not best practice to do on a template method. 
This sanity check can be done as e DEBUG assertion but I went with an an exception here since it's much easy to debug and workaround from JS if specific use case triggers it. Our tests should have caught this  long before it is hit by any user in a release version